File: src/main/java/com/kmatrokhin/uvbot/services/GeocodingService.java
package com.kmatrokhin.uvbot.services;

import com.fasterxml.jackson.databind.JsonNode;
import com.kmatrokhin.uvbot.dto.Coordinates;
import lombok.RequiredArgsConstructor;
import lombok.SneakyThrows;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;

@Service
@RequiredArgsConstructor
@Slf4j
public class GeocodingService {
    private static final String URL_TEMPLATE = "https://nominatim.openstreetmap.org/reverse?format=json&lat=%s&lon=%s&zoom=18&addressdetails=1&email=remillary@gmail.com";
    private final HttpExchangeService httpExchangeService;

    @SneakyThrows
    public String getLocationName(Coordinates coordinates) {
        log.info("Resolving location name for {}", coordinates);
        JsonNode jsonNode = httpExchangeService.request(URL_TEMPLATE, coordinates);
        String addressType = jsonNode.get("addresstype").asText();
        if (jsonNode.at("/address/city").isTextual()) {
            return jsonNode.at("/address/city").textValue();
        }
        return jsonNode.at("/address/" + addressType).textValue();
    }
}

File: src/main/java/com/kmatrokhin/uvbot/services/GeocodingService.kt
package com.kmatrokhin.uvbot.services

import com.kmatrokhin.uvbot.dto.Coordinates
import com.kmatrokhin.uvbot.dto.I18nProperties
import com.kmatrokhin.uvbot.entities.UserLanguage
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service

@Service
class GeocodingService(
    private val httpExchangeService: HttpExchangeService,
    private val i18nProperties: I18nProperties
) {
    private val log = LoggerFactory.getLogger(javaClass)

    fun getLocationName(coordinates: Coordinates): String =
        getLocationName(coordinates, UserLanguage.ENGLISH)

    fun getLocationName(coordinates: Coordinates, language: UserLanguage): String {
        log.info("Resolving location name for {}", coordinates)

        val acceptLanguage = when (language) {
            UserLanguage.RUSSIAN -> "ru,en"
            UserLanguage.ENGLISH -> "en,ru"
        }

        val jsonNode = httpExchangeService.request(
            "https://nominatim.openstreetmap.org/reverse" +
                "?format=jsonv2" +
                "&lat=${coordinates.latitude}" +
                "&lon=${coordinates.longitude}" +
                "&zoom=10" +
                "&addressdetails=1" +
                "&layer=address" +
                "&accept-language=$acceptLanguage" +
                "&email=konstant.matrokhin@gmail.com"
        )

        val error = jsonNode.path("error").asText("")
        if (error.isNotBlank()) {
            log.info("Nominatim reverse failed: {}", error)
            return i18nProperties.get(language, "unknown_place")
        }

        val address = jsonNode.path("address")
        val settlementKeys = listOf("city", "town", "village", "hamlet", "municipality", "locality")
        for (key in settlementKeys) {
            val value = address.path(key)
            if (value.isTextual && value.asText().isNotBlank()) {
                return value.asText()
            }
        }

        val displayName = jsonNode.path("display_name").asText("")
        val fallback = displayName.substringBefore(",").trim()
        return fallback.ifBlank { i18nProperties.get(language, "unknown_place") }
    }
}
